src/coreclr/jit/lower.cpp

```diff
@@ -1879,6 +1950,12 @@ void Lowering::LowerArg(GenTreeCall* call, CallArg* callArg, bool late)
         {
             ReplaceArgWithPutArgOrBitcast(ppArg, putArg);
         }
+
+        if (putArg->OperIsPutArgStk() && arg->OperIs(GT_LCL_VAR) &&
+            comp->lvaRequiresSpecialCopy(arg->AsLclVar()->GetLclNum()))
```

Comment:
This is the point where we should instead map back from the call + arg to the source local. Otherwise we are adding long term requirement that we end up with a `LCL_VAR` node here after all JIT transformations.

Reply:
I would probably move the handling to a separate function that is called right before `LegalizeArgPlacement` in `LowerArgsForCall`.

Would it be too crude to do it something like this:
```cpp
if (IsILStub() && call->IsPInvoke())
{
  unsigned argIndex = 0;
  for (CallArg& arg : call->gtArgs.Args())
  {
    if (!arg.IsUserArg())
    {
      continue;
    }

    if (argIndex >= comp->info.compArgsCount)
    {
      break;
    }

    // check if parameter at the exact same index as the IL argument is marked as requiring special copy, assuming that it is being passed 1:1 to the pinvoke
    if (lvaRequiresSpecialCopy(argIndex))
    {
      InsertCopyConstructor(arg); 
    }
  }
}
```
?